cedar-policy-validator/src/entity_manifest.rs

```diff
@@ -0,0 +1,1266 @@
+/// The keys to this map form the edges in the access trie,
+/// pointing to sub-tries.
+// CAUTION: this type is publicly exported in `cedar-policy`.
+// Don't make fields `pub`, don't make breaking changes, and use caution
+// when adding public methods.
+#[doc = include_str!("../../cedar-policy/experimental_warning.md")]
+pub type Fields<T> = HashMap<SmolStr, Box<AccessTrie<T>>>;
+
+/// The root of a data path or [`RootAccessTrie`].
+// CAUTION: this type is publicly exported in `cedar-policy`.
+// Don't make fields `pub`, don't make breaking changes, and use caution
+// when adding public methods.
+#[doc = include_str!("../../cedar-policy/experimental_warning.md")]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Hash)]
+pub enum EntityRoot {
+    /// Literal entity ids
+    Literal(EntityUID),
+    /// A Cedar variable
+    Var(Var),
+}
+
+impl Display for EntityRoot {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            EntityRoot::Literal(l) => write!(f, "{l}"),
+            EntityRoot::Var(v) => write!(f, "{v}"),
+        }
+    }
+}
+
+/// A [`RootAccessTrie`] is a trie describing a set of
+/// data paths to retrieve. Each edge in the trie
+/// is either a record or entity dereference.
+///
+/// If an entity or record field does not exist in the backing store,
+/// it is safe to stop loading data at that point.
+///
+/// `T` represents an optional type annotation on each
+/// node in the [`AccessTrie`].
+// CAUTION: this type is publicly exported in `cedar-policy`.
+// Don't make fields `pub`, don't make breaking changes, and use caution
+// when adding public methods.
+#[doc = include_str!("../../cedar-policy/experimental_warning.md")]
+#[serde_as]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct RootAccessTrie<T = ()>
+where
+    T: Clone,
+{
+    /// The data that needs to be loaded, organized by root.
+    #[serde_as(as = "Vec<(_, _)>")]
+    #[serde(bound(deserialize = "T: Default"))]
+    pub trie: HashMap<EntityRoot, AccessTrie<T>>,
+}
+
+/// A Trie representing a set of data paths to load,
+/// starting implicitly from a Cedar value.
+///
+/// `T` represents an optional type annotation on each
+/// node in the [`AccessTrie`].
+///
+// CAUTION: this type is publicly exported in `cedar-policy`.
+// Don't make fields `pub`, don't make breaking changes, and use caution
+// when adding public methods.
+#[doc = include_str!("../../cedar-policy/experimental_warning.md")]
+#[serde_as]
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct AccessTrie<T = ()> {
+    /// Child data of this entity slice.
+    /// The keys are edges in the trie pointing to sub-trie values.
+    #[serde_as(as = "Vec<(_, _)>")]
+    pub children: Fields<T>,
+    /// For entity types, this boolean may be `true`
+    /// to signal that all the ancestors in the entity hierarchy
+    /// are required (transitively).
+    pub ancestors_required: bool,
```

Comment:
depending on the use case for this serialization we'll probably want to include `#[serde(rename_all = "camelCase")]` on this and other structs. IIRC, all our external JSON formats use camel case. Not sure if we're consistent for internal DRT JSON